### DISCUSSION TITLE
Convergence of Berry curvature dipole near band crossing

### DISCUSSION THREAD

**Discussion by Bhuvaneswari-R432** [wannier-berri/wannier-berri#239] [Q&A]
Dear sir,

I am working on metallic systems and I want to calculate Berry curvature dipole. With either Fermi surface or Fermi sea method, variation in calculated BCD is huge near the band crossing region (between any two narrow bands) for different NK (I tried upto 520 x 520 x 1 grid, still the variation is significant). I also tried giving adpt_mesh and adpt_num_iter in run() calculation. Still, the BCD does not converge. Kindly suggest me ways to achieve convergence of BCD near band crossings.

**Comment by stepan-tsirkin** [comment]
Hi @Bhuvaneswari-R432 

BCD is really a tough quantity to integrate. I made a lot of affort to make it converge faster, but still one needs some shamanic dancing before it converges.  Bellow are some tips, I hope other users may add their experience

- Make sure you Wannier funcitpons are "good" (well localised, rather symmetric, etc)
- use symmetrized Wannier model https://wannier-berri.org/tutorials/symmetrization/Symmetrization_tutorial.html
- use symmetries (`system.set_symmetriues(...)`, `run(...,use_irred_kpt=True, symmetrize=True,...)` )
- Use tetrahedron method (`tetra=True` in the calculator)
- Start from a dense grid (as much as you can afford) and use a large adpt_num_iter
- adaptive refinement works more efficient when you have less calculators in one `run()`, and when the Fermi level range is narrow.  Isolate the problematic region and converge it. the other regions may be calculated separately, and then just concatinate arrays.
- As far as your system is 2D use ` System(..., periodic=[True,True,False],...)`
- In 2D 520 is not that much, it is only $\approx 2.5\cdot 10^{5}$ points. in 3D we sometimes use upto $10^8$ points
- If you start from wannier90, the Berry curvature is contributed by "internal" and "external" terms (on separation of terms see https://arxiv.org/abs/2303.10129 , eqs 62-66. Btw, if you use Fermi-sea evaluation, cite that paper, it describes the formalism behind the implementation) External terms are easier to converge, but heavier to calculate. You may switch them on/off by using  
 ```
system= wannierberri.system.System_w90( ... , use_wcc_phase=True)
calculators = {
"BD_int" :wannierberri.calculators.static.BerryDipole_FermiSea(.., tetra=True, kwargs_formula={'external_terms'=False} ),
"BD_ext" :wannierberri.calculators.static.BerryDipole_FermiSea(.., tetra=True, kwargs_formula={'internal_terms'=False} ),
}
```
see : 
https://github.com/wannier-berri/wannier-berri/blob/6b8f6bec42d1f689ca45c6019c33ea7f8fe558be/wannierberri/calculators/static.py#L21-L23
https://github.com/wannier-berri/wannier-berri/blob/6b8f6bec42d1f689ca45c6019c33ea7f8fe558be/wannierberri/formula/__init__.py#L6-L17
https://github.com/wannier-berri/wannier-berri/blob/6b8f6bec42d1f689ca45c6019c33ea7f8fe558be/wannierberri/formula/covariant.py#L125-L167

You may converge the external terms on a rather moderate grid, and then switch them off and go to denser grids with only the lighter (but problematic) internal terms. In most cases external terms give a smaller contribution then the internal ones.

- I am also experimenting with tetrahedron grids, but that is not tested nor documented. And anyway, I am not sure if ti will work well in 2D (see PR #233) .

**Reply by Bhuvaneswari-R432** [reply]
Thank you so much for your tips.

Regards,
Bhuvaneswari R
